File: ankor-websocket-connector/src/main/java/at/irian/ankor/switching/connector/websocket/WebSocketMessage.java
package at.irian.ankor.switching.connector.websocket;

import at.irian.ankor.action.Action;
import at.irian.ankor.change.Change;
import com.fasterxml.jackson.annotation.JsonInclude;

import java.util.Map;
import java.util.Set;

/**
 * @author Thomas Spiegl
 */
public class WebSocketMessage {
    @JsonInclude(JsonInclude.Include.NON_NULL)
    private String property;

    @JsonInclude(JsonInclude.Include.NON_NULL)
    private Action action;

    @JsonInclude(JsonInclude.Include.NON_NULL)
    private Change change;

    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Map<String,Object> stateValues;

    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Set<String> stateProps;

    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Map<String,Object> connectParams;

    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
    private boolean close;

    /**
     * for deserialization only
     */
    @SuppressWarnings("UnusedDeclaration")
    protected WebSocketMessage() {}

    private WebSocketMessage(String property,
                          Action action,
                          Change change,
                          Map<String, Object> stateValues,
                          Set<String> stateProps,
                          Map<String, Object> connectParams,
                          boolean close) {
        this.property = property;
        this.action = action;
        this.change = change;
        this.stateValues = stateValues;
        this.stateProps = stateProps;
        this.connectParams = connectParams;
        this.close = close;
    }

    public String getProperty() {
        return property;
    }

    public Action getAction() {
        return action;
    }

    public Change getChange() {
        return change;
    }

    public Map<String, Object> getStateValues() {
        return stateValues;
    }

    public Set<String> getStateProps() {
        return stateProps;
    }

    public Map<String, Object> getConnectParams() {
        return connectParams;
    }

    public boolean isClose() {
        return close;
    }

    public static WebSocketMessage createActionMsg(String property,
                                                   Action action,
                                                   Map<String, Object> stateValues,
                                                   Set<String> stateProps) {
        return new WebSocketMessage(property, action, null, stateValues, stateProps, null, false);
    }

    public static WebSocketMessage createChangeMsg(String property,
                                                   Change change,
                                                   Map<String, Object> stateValues,
                                                   Set<String> stateProps) {
        return new WebSocketMessage(property, null, change, stateValues, stateProps, null, false);
    }

    public static WebSocketMessage createConnectMsg(String modelName,
                                                    Map<String, Object> connectParameters) {
        return new WebSocketMessage(modelName, null, null, null, null, connectParameters, false);
    }

    public static WebSocketMessage createCloseMsg(String modelName) {
        return new WebSocketMessage(modelName, null, null, null, null, null, true);
    }

    @Override
    public String toString() {
        return "WebSocketMessage{" +
                "property='" + property + '\'' +
                ", action=" + action +
                ", change=" + change +
                ", connectParams=" + connectParams +
                ", close=" + close +
                '}';
    }
}

File: #ANKOR.net/Ankor.Core/Messaging/WebSockets/WebSocketMessage.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Ankor.Core.Action;

namespace Ankor.Core.Messaging.WebSockets {

	public class WebSocketMessage {
		
		public string Property { get; private set; }

		public AAction Action { get; private set; }

		public Change.Change Change { get; private set; }

		public IDictionary<string, object> ConnectParams { get; private set; }

		public bool? Close { get; private set; }

		private WebSocketMessage() {
			Close = null;
		}

		public static WebSocketMessage CreateConnectMsg(string modelName, IDictionary<string, object> connectParams) {
			return new WebSocketMessage { Property = modelName, ConnectParams = connectParams };			
		}

		public static WebSocketMessage CreateActionMsg(string property, AAction action) {
			return new WebSocketMessage {Action = action, Property = property };
		}

		public static WebSocketMessage CreateChangeMsg(string property, Change.Change change) {
			return new WebSocketMessage { Change = change, Property = property };
		}

		public static WebSocketMessage CreateCloseMsg(string modelName) {
			return new WebSocketMessage { Close = true, Property = modelName };
		}

	}
}
